### ISSUE TITLE
Error when setting <sessionindicator> on data where runs ~ sessions

### ISSUE THREAD

**Issue by oliver-contier** [cvnlab/GLMsingle#8]
Not sure whether this is a bug or my fault for incorrectly handling the code.

As you know, the THINGS dataset has 100 repeated stimuli in each session. Hence, I've concatenated all runs in a given session. Basically, I'm pretending that our sessions are runs...

This worked fine with default settings, but an exception is thrown from within `calcbadness()` when I specify `sessionindicator` in the `opt`.

```
nses = 3 # testing on three sessions

gs = GLM_single(params=dict(
    wantlibrary=1, wantglmdenoise=1, wantfracridge=1,
    wantfileoutputs=[1,1,1,1],
    sessionindicator=list(range(nses)),
))

# data is list of length 3, each element represents one session (all runs concatenated -> array shape (72, 91, 1, 2840)). 
# design is corresponding list of length 3, each element is a matrix of shape (2840, 3080)

results_glmsingle = gs.fit(design, data, stimdur=.5, tr=1.5, 
                           outputdir=pjoin(outdir, 'glmsingle'))
```
```
*** FITTING TYPE-A MODEL (ONOFF) ***
fitting model...
done.
preparing output...
done.
computing model fits...
done.
computing R^2...
chunks:   0%|          | 0/1 [00:00<?, ?it/s]
done.
computing SNR...
done.
*** Saving results to /LOCAL/ocontier/thingsmri/bids/code/sandbox/test_glmsingle/glmsingle/TYPEA_ONOFF.npy. ***
*** Setting brain R2 threshold to 0.09844616397985788 ***
*** FITTING TYPE-B MODEL (FITHRF) ***
chunks: 100%|██████████| 1/1 [36:20<00:00, 2180.27s/it]
*** Saving results to /LOCAL/ocontier/thingsmri/bids/code/sandbox/test_glmsingle/glmsingle/TYPEB_FITHRF.npy. ***
*** DETERMINING GLMDENOISE REGRESSORS ***
chunks:   0%|          | 0/1 [00:00<?, ?it/s]
*** CROSS-VALIDATING DIFFERENT NUMBERS OF REGRESSORS ***
chunks:   0%|          | 0/1 [19:14<?, ?it/s]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-22-d4954029f960> in <module>
      6 ))
      7 
----> 8 results_glmsingle = gs.fit(design, data_concat, stimdur=.5, tr=tr, 
      9                            outputdir=pjoin(outdir, 'glmsingle'))
/LOCAL/ocontier/thingsmri/bids/code/external_libraries/GLMsingle/glmsingle/glmsingle.py in fit(self, design, data, stimdur, tr, outputdir)
   1017 
   1018                         results0.append(temp['betasmd'])
-> 1019                     glmbadness[relix, :] = calcbadness(
   1020                         params['xvalscheme'],
   1021                         validcolumns,
/LOCAL/ocontier/thingsmri/bids/code/external_libraries/GLMsingle/glmsingle/ssq/calcbadness.py in calcbadness(xvals, validcolumns, stimix, results, sessionindicator)
     71         wh = np.flatnonzero(sessionindicator == sess)
     72 
---> 73         whcol = np.concatenate(np.asarray(validcolumns)[wh])
     74 
     75         # mean of unregularized case
<__array_function__ internals> in concatenate(*args, **kwargs)
ValueError: need at least one array to concatenate
```

Cheers!
Oli

**Comment by kendrickkay** [comment]
@oliver-contier -> i would recommend to not concatenate the runs within a given scan session. One reason is that the way the polynomials work is to assume that the time-series are continuous within a run. But once you concatentate, that assumption is false.  Easier to just give GLMestimatesingletrial.m all of your data as individual runs (and from multiple scan sessions)

Regarding the error, I believe it's due to the use of 0,1,2.  The code wants positive integers.  Or at least, that's what the MATLAB version required and I assume Ian kept that requirement.   So give it something like [1,1,1,1,1,1,2,2,2,2,2,3,3,3,3,3,3] for your case of aggregating runs across 3 scan sessions?

**Comment by kendrickkay** [comment]
I did more detective work. I think there may be bugs.

`In [15]: opt['sessionindicator']
Out[15]: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 2, 1]

In [16]: np.max(opt['sessionindicator'])
Out[16]: 10

In [17]: sessions=range(1,10)

In [18]: sessions
Out[18]: range(1, 10)

In [21]: for sess in sessions:
    ...:     print(f'{sess}')
    ...: 
1
2
3
4
5
6
7
8
9

In [22]: sess=1

In [23]: opt['sessionindicator']==sess
Out[23]: False

In [24]: np.flatnonzero(opt['sessionindicator']==sess)
Out[24]: array([], dtype=int64)
`

In the above, notice that the intention is that sessionindicator contains positive integers.  In this case, 1 through 10.  The range is from 1 to 10, which apparently is interpreted as just 1 through 9.  Also, the line (opt['sessionindicator']==sess) does not seem to behave like Matlab, as it just returns a logical.

@iancharest can you check over this and fix accordingly?

**Comment by oliver-contier** [comment]
Thanks for the explanation @kendrickkay, I'll make sure not to concatenate runs for each session.

And thanks for the detective work! Regarding `sessionindicator` and `xvalscheme`, I'm not quite sure now whether to use indexing starting at 0 (which is default in python) or 1 (which is default in Matlab, I assume, and also used in the documentation examples). Might there be a difference between the two implementations?

**Comment by kendrickkay** [comment]
i believe that jacob or ian needs to implement the fix and then you can go ahead.  i think the desired behavior should match Matlab and therefore be positive integers only.

**Comment by oliver-contier** [comment]
OK, thanks!

**Comment by kendrickkay** [comment]
of course, you are welcome to try to fix it yourself and open a pull request.  (I personally am very impatient and hate to be bottlenecked!)   the problem here is, i don't really know how to speak Python

**Comment by jacob-prince** [comment]
oliver, i will submit a pull request in the next hour to implement the fixes kendrick alluded to above. at that point, the code should run without error if the user provides a sessionindicator. however, i have not finished checking methodically whether the python and matlab outputs match one-another when sessionindicator is used. so proceed with that caveat in mind.

**Comment by jacob-prince** [comment]
this issue has been resolved to the extent that sessionindicator no longer throws errors, but i will keep the issue open until i've been able to verify that the outputs of the matlab and python versions are identical given the same inputs.

**Comment by kendrickkay** [comment]
that's probably a good idea.  err on the side of paranoia, i suppose.  good practice.

**Comment by kendrickkay** [comment]
Looks like this was fixed a long time ago. Merged and closed with pull request #11 .
